cleverhans/discretization/attacks.py

```diff
@@ -0,0 +1,246 @@
+"""Contains all the attacks on discretized inputs.
+
+The attacks implemented are Discrete Gradient Ascent (DGA) and
+Logit Space-Projected Gradient Ascent (LS-PGA).
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import numpy as np
+import tensorflow as tf
+import discretization_utils
+
+
+def discretize_range(discretize_fn, levels, low, high, thermometer=False):
+  """Get range of discretized values in the interval (low, high).
+
+    For example, assume discretize_fn uniformly discretizes the values
+    between 0 and 1 into 10 bins each represented by either a one hot encoding
+    or a thermometer encoding. Then discretize_range(discretize_fn, .3, .7)
+    would return [0., 0., 0., 1., 1., 1., 1., 0., 0., 0.]. Note that it's output
+    is independent of the encoding used.
+
+  Args:
+    discretize_fn: Discretization function used to discretize input.
+    levels: Number of levels to discretize the input into.
+    low: Minimum value in the interval.
+    high: Maximum value in the interval.
+    thermometer: If True, then the discretize_fn returns thermometer codes,
+        else it returns one hot codes. (Default: False).
+
+  Returns:
+    Mask of 1's over the interval.
+  """
+  low = tf.clip_by_value(low, 0., 1.)
+  high = tf.clip_by_value(high, 0., 1.)
+  out = 0.
+  for alpha in np.linspace(0., 1., levels):
+    q = discretize_fn(alpha * low + (1. - alpha) * high, levels, thermometer)
+
+    # Convert into one hot encoding if q is in thermometer encoding
+    if thermometer:
+      q = discretization_utils.thermometer_to_one_hot(q, levels, flattened=True)
+    out += q
+  return tf.to_float(tf.greater(out, 0.))
+
+
+def adv_dga(x, model, discretize_fn, projection_fn, levels, phase,
+			steps, eps, thermometer=False, noisy_grads=True, y=None):
```

Comment:
This function doesn't ever seem to be called anywhere. Is it just included for completeness because we describe it in the paper as one of the weaker attacks we tried? If so (If we're just including it as a reference attack implementation) it should be an Attack subclass. If not, maybe we should delete it.

Reply:
If we keep this function, it should have at least one test

Reply:
It's used if you want to actually discretize the inputs before calling a model.